csharp/src/Ice/Reference.cs

```diff
@@ -22,148 +22,251 @@ internal interface IGetConnectionCallback
             }
+            if (Protocol != other.Protocol)
+            {
+                return false;
+            }
+
             return true;
         }
 
-        //
-        // Marshal the reference.
-        //
-        public virtual void StreamWrite(OutputStream s)
+        public override int GetHashCode()
         {
-            //
-            // Don't write the identity here. Operations calling streamWrite
-            // write the identity.
-            //
-
-            //
-            // For compatibility with the old FacetPath.
-            //
-            if (_facet.Length == 0)
+            if (_hashCode != 0)
             {
-                s.WriteStringSeq(Array.Empty<string>());
+                // Already computed, return cached value:
+                return _hashCode;
             }
             else
             {
-                string[] facetPath = { _facet };
-                s.WriteStringSeq(facetPath);
-            }
-
-            s.WriteByte((byte)InvocationMode);
-
-            s.WriteBool(IsSecure);
+                // Lazy initialization of _hashCode to a value other than 0. Reading/writing _hashCode is atomic.
+                var hash = new HashCode();
 
-            s.WriteByte((byte)_protocol);
-            s.WriteByte(0);
-            s.WriteByte(_encoding.Major);
-            s.WriteByte(_encoding.Minor);
+                // common properties
+                if (Compress != null)
+                {
+                    hash.Add(Compress.Value);
+                }
+                hash.Add(Collections.GetHashCode(Context));
+                hash.Add(Encoding);
+                hash.Add(Facet);
+                hash.Add(Identity);
+                hash.Add(InvocationMode);
+                hash.Add(IsSecure);
+                hash.Add(Protocol);
+
+                if (IsFixed)
+                {
+                    hash.Add(_fixedConnection);
+                }
+                else
+                {
+                    hash.Add(AdapterId);
+                    hash.Add(ConnectionId);
+                    hash.Add(ConnectionTimeout);
+                    hash.Add(Collections.GetHashCode(Endpoints));
+                    hash.Add(IsCollocationOptimized);
+                    hash.Add(IsConnectionCached);
+                    hash.Add(IsPreferSecure);
+                    hash.Add(EndpointSelection);
+                    hash.Add(LocatorCacheTimeout);
+                    if (LocatorInfo != null)
+                    {
+                        hash.Add(LocatorInfo);
+                    }
+                    if (RouterInfo != null)
+                    {
+                        hash.Add(RouterInfo);
+                    }
+                }
 
-            // Derived class writes the remainder of the reference.
+                int hashCode = hash.ToHashCode();
+                if (hashCode == 0)
+                {
+                    // 0 means uninitialized so we switch to 1
```

Comment:
Why not use a nullable here instead?

Reply:
This code/logic did not change. The concern with a nullable is read ordering. Say `_hashCode` would be nullable,  would:
```
if (_hashCode != null)
{
    return _hashCode.Value;
}
```
always read the initialized `_hashCode` value? (Remember, there is no synchronization).